### DISCUSSION TITLE
add_configfiles 的使用疑问

### DISCUSSION THREAD

**Discussion by BanKnight** [xmake-io/xmake#2239] [Q&A] [answered]
我今天测试 add_configfiles  有些疑问

xmake.lua
```lua
add_rules("mode.debug", "mode.release")
set_configvar("MYVERSION", 22)
target("test_xmake")
    set_kind("binary")
    add_files("src/*.cpp")
    add_configfiles("src/version.h.in", {pattern = "@(.-)@"})
```
make.cpp
```cpp
#include <iostream>
#include "version.h"

using namespace std;

int main(int argc, char** argv)
{
    cout << "hello world!" << endl;
    return 0;
}
```
version.h.in
```cpp
#define  MYVERSION @MYVERSION@
```
发现可以成功生成 build/version.h 文件，但是编译失败，提示：
```
$ xmake
generating src/version.h.in ... ok
[ 25%]: compiling.release src/main.cpp
error: src/main.cpp:2:10: fatal error: version.h: No such file or directory
    2 | #include "version.h"
      |          ^~~~~~~~~~~
compilation terminated.
  > in src/main.cpp
```
## 疑问
add_configfiles 生成的文件不会自动加入到include文件列表吗？如果设定如此，那么应该如何达成这样的目的？

**Comment by waruqi** [comment, marked as answer]
看下文档，有详细说明，指定下输出目录

https://xmake.io/#/zh-cn/manual/project_target?id=targetadd_configfiles

**Comment by BanKnight** [comment]
好的，可以设定目录后，的确可以解决问题。

我刚刚又修改了一下version.h.in，增加了一行用来测试pattern，疑似有bug
version.h.in
```cpp
//@mail.163 ，这里是新增的行，不应该按照pattern替换
#define  MYVERSION @MYVERSION@
```
执行报错
```
$ xmake
error: cannot get variable(mail.163 ，这里是新增的行，不应该按照pattern替换
#define  MYVERSION) in src/version.h.in.
```
这里错误地将 @mail.163 按照pattern解释了

**Reply by waruqi** [reply]
既然你用了 `@@`，说明你自己指定的 pattern，那就自己调整下 pattern 就行了，处理下换行，`pattern = "@([^\n]-)@"`

**Comment by BanKnight** [comment]
我理解是，它是匹配到了 @mail.163，但是理应匹配不出来才对，我用正则测试工具去测试，也是匹配不出来的。

**Reply by waruqi** [reply]
> 我理解是，它是匹配到了 @mail.163，但是理应匹配不出来才对，我用正则测试工具去测试，也是匹配不出来的。

你改成 `pattern = "@([^\n]-)@"` 试下就知道了，lua 模式匹配，不是正则

**Comment by BanKnight** [comment]
改了之后的确可以成功生成了。

然后刚刚我又做了一个新的测试，我主动将生成的 version.h 文件删掉，
然后再 xmake，发现它不会重新生成。

**Reply by waruqi** [reply]
执行下 xmake f -c ，这个只在配置阶段生成
